### PR TITLE
Add CORS setting and addNote audio downloading support

### PR DESCRIPTION
**Pull request by sp3ctum**
# Adds a new setting called `cors_host`
This setting, when set, enables CORS (cross origin resource sharing) headers in the API. These allow browser access from the configured site. By default, browsers are denied access, and only when CORS headers are applied is the access given.

The new setting can be configured in the settings:

image

Related issue: https://github.com/KamWithK/AnkiconnectAndroid/issues/13

# Adds support for the addNote api to download audio files from given URLs
An example of such a request can be seen here: https://github.com/KamWithK/AnkiconnectAndroid/issues/17

### CODE REVIEW

File: app/src/main/java/com/kamwithk/ankiconnectandroid/routing/RouteHandler.java
Comment (Aquafina-water-bottle) [inline app/src/main/java/com/kamwithk/ankiconnectandroid/routing/RouteHandler.java:66]: Can you make sure this doesn't inadvertently affect Forvo audio? If memory serves me correct, I tried to add this in a different PR, but removed it because the Forvo audio was no longer getting fetched.

P.S. Thanks for the PR! I'll try to make some time to review it properly sometime in the next few days.

Comment (sp3ctum) [reply app/src/main/java/com/kamwithk/ankiconnectandroid/routing/RouteHandler.java:66]: Sure! I'll set up my yomichan according to the instructions in the readme.

No rush with the PR, take your time!

Comment (sp3ctum) [reply app/src/main/java/com/kamwithk/ankiconnectandroid/routing/RouteHandler.java]: You were right - forvo audio was broken by this change. I think I was able to fix it, though: see 87c2ecd7c034e66734915b2ea3d76c01da671128

In yomichan, I can now select all these audios and hear the word being read by different people
image

Comment (Aquafina-water-bottle) [reply app/src/main/java/com/kamwithk/ankiconnectandroid/routing/RouteHandler.java]: Looks good, thanks for the change! I'll temporarily keep this open as a self reminder to verify that it works on my end.

**Comment by KamWithK** [comment]
Thanks for the contribution, I'll take a look when I get some time!

File: app/src/main/java/com/kamwithk/ankiconnectandroid/routing/RouteHandler.java
Comment (sp3ctum) [reply app/src/main/java/com/kamwithk/ankiconnectandroid/routing/RouteHandler.java]: Just noticed I had accidentally broken my cors support with the fix. :)
Pushed some more changes - now forvo works locally and the cors headers are also applied correctly.

File: app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/IntegratedAPI.java
Comment (Aquafina-water-bottle) [inline app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/IntegratedAPI.java:98]: I believe these changes should be completely moved to `MediaAPI`. My understanding of `IntegratedAPI` is that this class is only used when:
- More than one API is being used (e.g. `addNote`)
- The Android `Context` is needed (e.g. `guiBrowse`)

File: app/src/main/java/com/kamwithk/ankiconnectandroid/routing/AnkiAPIRouting.java
Comment (Aquafina-water-bottle) [inline app/src/main/java/com/kamwithk/ankiconnectandroid/routing/AnkiAPIRouting.java:215]: Fortunately, these changes still work on images (I tested this on a Yomichan dictionary with images, and it exported perfectly fine). However, changing this is a misnomer, because the media can be images or audio (and theoretically, anything other byte array). I think it would be safe to undo the logic, or rename the `AudioFile` class to something like `BinaryFile`.

File: app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/MediaAPI.java
Comment (Aquafina-water-bottle) [inline app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/MediaAPI.java:39]: Just curious about this: did adding this logic solve something on your end?

**Review by Aquafina-water-bottle — CHANGES_REQUESTED** [review]
Overall, everything looks good! Especially appreciate the documentation in the code. There were just a few nitpicks and questions.

Note that I didn't actually test the main parts of this PR (downloading audio from `addNote` and the CORS additions), but I tried running it through the standard Yomichan setup, and tested everything else I think it could've affected, and everything there seems to work fine for me. (We might want to eventually add some tests to this, but that's outside the scope of this PR.)

P.S. Thank you for getting the `Ankiconnect Android is running.` message working! It'll be very useful to debug people's setups, and once this gets merged, I'll make sure to add it somewhere in the README.

File: app/src/main/java/com/kamwithk/ankiconnectandroid/routing/RouteHandler.java
Comment (Aquafina-water-bottle) [reply app/src/main/java/com/kamwithk/ankiconnectandroid/routing/RouteHandler.java]: Seems like everything here works, thanks!

**Comment by Aquafina-water-bottle** [comment]
The changes look good, thanks! One last thing, as mentioned in #31, we're planning on quickly creating a `docs/api.md` file, where edge cases such `addNote` not supporting downloaded files can be properly documented. I'm thinking it would be best to block merging this PR until that file gets added to master, and until you can add this call into the file.

**Comment by KamWithK** [comment]
In agreement with @Aquafina-water-bottle about blocking this until docs are added

**Comment by sp3ctum** [comment]
Sure! I subscribed to that PR and will add documentation after it gets merged.

**Comment by KamWithK** [comment]
Now that the documentation PR has been merged, add some docs and then once you're ready we can take another look at this PR @sp3ctum

**Comment by sp3ctum** [comment]
Just added (a little) documentation and squashed the commits for a cleaner history. I think this is ready for another review round whenever you have the time.

**Comment by Aquafina-water-bottle** [comment]
Thanks for the documentation, everything looks good to me!

**Comment by KamWithK** [comment]
Looks good thanks @sp3ctum @Aquafina-water-bottle

**Pull request merged**
